content/blog/2023-12-11-array-ir/index.md

```diff
@@ -0,0 +1,109 @@
+
+The pre-circuit IR represents functions as a sequence of control flow statements and let-bindings. Circuits are not present in this IR. 
+
+The pre-circuit IR is different from the circuit IR in that:
+1. All let-bindings are annotated with a protocol. In the circuit IR, only let-bindings involving a “command” (input, output, and function/circuit calls) are annotated with protocols denoting where results are stored. Let-bindings involving computations occur within circuits, which are annotated with the associated protocol for all included statements. 
+2. It includes downgrade nodes. These are explicit violations of secure information flow, such as declassifying secret data or endorsing untrusted values. It is necessary that downgrade instructions are preserved in the precircuit IR, since they enforce a secure ordering on computations.
+
+# Grouping
+The grouping procedure places statements of the same protocol together whenever possible. This will enable larger circuit blocks, reducing the overhead of communication between protocols.
+
+This reordering is subject to data and security dependencies. Node `x` is a dependency of node `y` when:
+- `x` defines data that is used by `y`,
+- `x` is an `input` statement, and `y` an output,
+- `x` is an `endorse` statement, and `y` an declassify, or
+- `x` is an `output` statement, and `y` an input from the same host.
+
+The first rule corresponds to data dependencies. The second and third rules hold because moving a statement which reveals information before one that commits to data violates security. The last rule preserves the user interface.
+
+We decided to only reorder at top-level within blocks; that is, statements only get swapped if they exist at the same “depth.” For instance, consider the following code:
+```
+<A>
+if (condition) {
+	<B>
+	<C>
+}
+```
+Then `B` might be reordered with `C`, and the entire `if` block could be reordered with `A`. However, we choose not to allow reorderings where `B` or `C` may be lifted out of the block. This allows us to essentially ignore control dependencies, making reordering simpler.
+
+The grouping procedure constructs a dependency graph for each block (a sequence of code enclosed with curly braces), where nodes are statements. For instance, given the above code, the procedure constructs a dependency graph containing `A` and the `if` statement, and another for `B` and `C`. Reordering occurs only within these two graphs. Notice that these graphs do not need to maintain control-flow information, so statements can be reordered strictly based on data and security dependencies.
+
+
+We formalize the reordering task as the following problem. Given a directed acyclic graph $G = (V,E)$ where vertices are statements and edges are dependencies, and an assignment of protocols to nodes $p : V \to \mathbb{N}$, we aim to find a sorting $\pi$ of the nodes ($v_{\pi_1}, v_{\pi_2}, \dots, v_{\pi_n}$) such that:
```

Comment:
I think you need to add `latex = true` to the front matter to make the LaTeX math render.